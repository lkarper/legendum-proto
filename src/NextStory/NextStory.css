.NextStory__link {
    font-size: 20px;
}

.NextStory__link {
    color: #FFC71F;
    text-decoration: none;
    font-family: 'EB Garamond', serif;
    margin: 5px 10px;
    padding: 5px;
}

.NextStory__link:active {
    text-decoration: underline;
}

.NextStory__link:hover {
    color: goldenrod;
}