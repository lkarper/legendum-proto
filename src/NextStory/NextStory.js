import React, { useContext } from 'react';
import { Link } from 'react-router-dom'
import UserContext from '../contexts/UserContext';
import './NextStory.css';

const NextStory = (props) => {
    const context = useContext(UserContext);

    const exercises = [...context.exercises].sort((a, b) => a.chapter_number - b.chapter_number);

    const lastChapter = context.exercises.length !==0 
        ? [...context.exercises].sort((a, b) => b.chapter_number - a.chapter_number)[0].chapter_number 
        : undefined;

    let nextChapter;

    // Look for the first chapter that the user has not yet completed, and set it to be the next chapter
    if (context.progress.length !== 0) {
        for (const exercise of exercises) {
            if (context.progress.filter(p => p.chapter_number === exercise.chapter_number).length === 0) {
                nextChapter = exercise.chapter_number;
                break;
            }
        }
    }

    if (nextChapter) {
        return (
            <Link 
                className='NextStory__link button' 
                to={`/game/story/${nextChapter}`}
            >
                Pick up where you left off (chapter {nextChapter})
            </Link>
        );
    } else if (context.progress.length === 0) {
        return (
            <Link 
                className='NextStory__link button' 
                to={`/game/story/1`}
            >
                Start at Chapter 1
            </Link>
        );
    }
    
    return (
        <Link 
            className='NextStory__link button' 
            to={`/game/story/${lastChapter}`}
        >
            Visit the last chapter again
        </Link>
    );
}

export default NextStory;
