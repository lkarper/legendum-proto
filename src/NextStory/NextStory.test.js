import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import NextStory from './NextStory';

describe('NextStory component', () => {
  it('renders the UI without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <NextStory />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const wrapper = mount(
      <BrowserRouter>
        <NextStory />
      </BrowserRouter>
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
