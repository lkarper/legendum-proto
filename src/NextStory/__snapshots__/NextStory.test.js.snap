// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextStory component renders the UI as expected 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <NextStory>
      <Link
        className="NextStory__link"
        to="/game/story/1"
      >
        <LinkAnchor
          className="NextStory__link"
          href="/game/story/1"
          navigate={[Function]}
        >
          <a
            className="NextStory__link"
            href="/game/story/1"
            onClick={[Function]}
          >
            Start at Chapter 1
          </a>
        </LinkAnchor>
      </Link>
    </NextStory>
  </Router>
</BrowserRouter>
`;
